### PR TITLE
🛡️ feat: Persist & enforce `disable-model-invocation` / `user-invocable` / `allowed-tools`

### PR DESCRIPTION
**Pull request by danny-avila**
## Summary

Phase 6 of the [Agent Skills rollout](https://github.com/danny-avila/LibreChat/pull/12625) — wires the three already-parsed-and-validated frontmatter fields into actual runtime behavior. Up until now `user-invocable`, `disable-model-invocation`, and `allowed-tools` were accepted in `SKILL.md` and validated, but had **zero runtime effect**. This PR persists them as first-class columns derived from frontmatter and enforces them at every invocation path.

Targets `feat/agent-skills` (umbrella PR [#12625](https://github.com/danny-avila/LibreChat/pull/12625)). Builds on the precursor [#12744](https://github.com/danny-avila/LibreChat/pull/12744) which plumbed `allowedTools` through `resolveManualSkills` without consumption.

### Schema + types

Three new columns on `Skill` (and matching fields on `ISkill` / `TSkill` / `TSkillSummary`):

| Column | Default | Purpose |
|---|---|---|
| `disableModelInvocation` | `false` (indexed) | Excludes the skill from the catalog AND rejects model-side `skill` tool calls. |
| `userInvocable` | `true` | Hides from the `\$` popover AND skips manual invocation in the resolver. |
| `allowedTools` | `undefined` | Tool allowlist unioned into the agent's effective set when the skill is invoked. |

A `deriveStructuredFrontmatterFields` helper maps frontmatter → columns at create/update time. On update, fields the new frontmatter omits are `\$unset` so removing `disable-model-invocation` from a SKILL.md re-enables model invocation on save.

### Runtime enforcement (the four-quadrant matrix from the brief)

| `disable-model-invocation` | `user-invocable` | Catalog inject? | `\$` popover? | Model can `skill` tool call? |
|---|---|---|---|---|
| `false` (default) | `true` (default) | ✅ | ✅ | ✅ |
| `true` | `true` | ❌ | ✅ | ❌ |
| `false` | `false` | ✅ | ❌ | ✅ |
| `true` | `false` | ❌ | ❌ | ❌ (effectively disabled) |

- `injectSkillCatalog` filters out `disableModelInvocation: true` before formatting — they cost zero context tokens.
- `handleSkillToolCall` rejects with a clear error when the model names a disabled skill (defense against stale-cache / hallucinated invocation slipping past the catalog filter).
- `resolveManualSkills` skips `userInvocable: false` skills with a warn log so an API-direct caller can't bypass the popover-side filter.
- `SkillsCommand` popover replaces the phase-1 UI-only `invocationMode` check with the persisted `userInvocable !== false`.

### `allowed-tools` plumbing

A new `unionPrimeAllowedTools` helper folds skill-declared tool names into the union, minus what's already on the agent. `initialize.ts` re-runs `loadTools` for the extras and merges the resulting `toolDefinitions` into the agent's effective set **for the turn only**.

Behavior is **tolerant of unknown tool names**: the registry intersection silently drops anything that doesn't load, with a debug log attributing each missing name to the skill that requested it. This is the load-bearing requirement for cross-ecosystem skills — a Claude Code skill referencing `edit_file` (which LibreChat hasn't shipped yet) imports today and lights up automatically once the tool lands. No skill-side migration ever needed.

The agent document is **never modified** — the union is turn-scoped only.

### Phase 5 coordination

`unionPrimeAllowedTools` is structured as a pure helper that takes any `{ name, allowedTools? }[]` array. Phase 5's `alwaysApply` resolver passes `[...manualPrimes, ...alwaysApplyPrimes]` through the same helper — no second tool-loading code path.

## What's deliberately out of scope

- **Form UI for editing the new fields.** The `Skills/forms/*` components still render the phase-1 `invocationMode` picker; it's deprecated but functional (defaults to `auto`, discarded on save). Replacing it with proper `disable-model-invocation` / `user-invocable` toggles is a UX change with localization implications and belongs in a follow-up.
- **Phase 7's "unrecognized tool" import-preview warning.** The brief flags this as Phase 7's territory; the runtime drop-with-debug-log here is sufficient for v1 enforcement.
- **`allowed-tools` UX layer notice.** Surfaced in Phase 7's import preview UI per the brief.

## Test plan

- [x] `packages/data-schemas`: `deriveStructuredFrontmatterFields` unit tests cover string vs array `allowed-tools`, missing-field defaults, all-three-fields combo, non-boolean rejection.
- [x] `packages/data-schemas`: CRUD round-trip — create persists derived columns, update re-derives (and unsets) when frontmatter changes, list projection includes them.
- [x] `packages/api`: `injectSkillCatalog` excludes `disableModelInvocation: true` (regardless of active state).
- [x] `packages/api`: `resolveManualSkills` silently skips `userInvocable: false`, preserves the rest of the batch.
- [x] `packages/api`: skill tool handler rejects `disableModelInvocation: true` with a distinct error message; default-undefined skills go through.
- [x] `packages/api`: `unionPrimeAllowedTools` dedupes across skills, skips agent-baseline tools, filters non-string entries, preserves first-occurrence order, attributes per-skill.
- [x] `client`: `SkillsCommand` `userInvocable !== false` filter; existing per-agent / per-user / invocability layered tests still pass.
- [x] All affected suites green (`packages/api`, `packages/data-schemas`, `client`, legacy `api/server/routes/skills.test.js`, agent controller specs).
- [ ] Manual smoke: import a skill with `user-invocable: false`, verify it disappears from `\$` popover and stays in catalog.
- [ ] Manual smoke: import a skill with `disable-model-invocation: true`, verify catalog excludes it and skill tool call is rejected.
- [ ] Manual smoke: import a skill with `allowed-tools: ['execute_code']`, verify the model gains access to `execute_code` for the turn the skill is invoked.

## Notes for reviewers

- **Phase 5 is in flight in parallel** by another contributor; touch points limited to `unionPrimeAllowedTools` (Phase 5 will pass both prime arrays through it) and `ResolvedManualSkill.allowedTools` (already landed via #12744).
- **Pre-existing client `tsc` errors** in `Capabilities` / `Mermaid` / `LiveMessage` files are unchanged by this PR (verified via `git stash` + `tsc` rerun on the base commit).
- **`getPublicSkillIdSet` lint warning** in `packages/api/src/skills/handlers.ts` is pre-existing — out of scope for this PR.

### CODE REVIEW

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/initialize.ts [hunk 1]
```diff
@@ -526,6 +551,86 @@ export async function initializeAgent(
...
+          provider,
+          agentId: agent.id,
+          tools: extraToolNames,
+          model: agent.model,
+          tool_options: agent.tool_options,
+          tool_resources,
+        });
+        const loadedDefs = extraResult?.toolDefinitions ?? [];
+        const loadedNames = new Set(loadedDefs.map((d) => d.name));
+        const droppedNames = extraToolNames.filter((n) => !loadedNames.has(n));
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/initialize.ts]: Addressed in fbd002151:

- **P2 (MCP auth context)** — Restructured to a single `loadTools` call. `manualSkillPrimes` resolution moved before tool-load so `agent.tools + allowedTools` go through one pass; `userMCPAuthMap` / `toolContextMap` / `actionsEnabled` for skill-added MCP tools now flow through to `InitializedAgent` and downstream `loadToolsForExecution`. Also wrapped in try/catch so a bad `allowed-tools` entry retries with just `agent.tools` instead of crashing the turn.

- **P3 (catalog filter shadows handler reject)** — `injectSkillCatalog.activeSkillIds` now keeps `disable-model-invocation: true` skills so `handleSkillToolCall` can fetch the doc and fire the explicit "cannot be invoked by the model" rejection. Catalog text and tool registration still gate on the visible subset (zero context tokens preserved); per-user `isActive` is the only hard exclusion now.

Six new integration tests in `initialize.test.ts` cover the union pass-through, retry-on-throw, and edge cases. Two `injectSkillCatalog` tests updated for the new `activeSkillIds` semantics.

File: packages/api/src/agents/skills.ts [hunk 2]
```diff
@@ -196,6 +196,16 @@ export async function injectSkillCatalog(
...
       }
+      /**
+       * `disable-model-invocation: true` skills cost zero context tokens
+       * when not in use — they're filtered out before catalog formatting
+       * so the model never sees them in its system prompt and can't name
+       * them via the `skill` tool. Manual `$` invocation paths bypass this
+       * filter; they're gated separately in `resolveManualSkills`.
+       */
+      if (skill.disableModelInvocation === true) {
+        continue;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts]: Addressed in fbd002151. `injectSkillCatalog` now collects all active skills (regardless of `disable-model-invocation`) into `activeSkillIds`, then filters the catalog text + tool registration on the visible subset only. Result:

- Catalog text excludes `disable-model-invocation: true` skills (zero context tokens — unchanged).
- Tool registration still skips entirely if every active skill is model-disabled (no point registering the skill tool with nothing to call).
- `activeSkillIds` (which becomes the runtime `accessibleSkillIds` for `getSkillByName`) keeps disabled skills, so a stale/hallucinated model invocation now reaches `handleSkillToolCall` and gets the explicit "cannot be invoked by the model" rejection instead of the misleading generic "not found".

Two existing tests updated for the new semantics; the previous one asserting `activeSkillIds` excluded the disabled skill now asserts the inverse and explains why.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/skills.ts [hunk 3]
```diff
@@ -219,9 +231,26 @@ export async function injectSkillCatalog(
...
   }
 
-  // Warn on duplicate names — model may invoke the wrong skill
+  /**
+   * Catalog text and tool registration are gated on the *visible* subset.
+   * If every active skill is `disable-model-invocation: true`, the model
+   * can't reach any of them anyway — registering the skill tool would
+   * burn context tokens for nothing.
+   */
+  const catalogVisibleSkills = activeSkills.filter((s) => s.disableModelInvocation !== true);
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts:253]: Addressed in ea09553de. The catalog cap now counts only model-visible skills (`visibleCount` tracks only `disableModelInvocation !== true`), so disabled rows no longer consume quota slots. `MAX_CATALOG_PAGES` stays the scan-budget ceiling. New test asserts that 3 leading disabled rows + 1 trailing visible skill correctly produce a catalog with the visible one.

File: packages/api/src/agents/skills.ts [hunk 4]
```diff
@@ -277,7 +306,7 @@ export async function injectSkillCatalog(
 
   return {
     toolDefinitions,
-    skillCount: activeSkills.length,
+    skillCount: catalogVisibleSkills.length,
     activeSkillIds: activeSkills.map((s) => s._id),
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts]: Addressed in ea09553de. New `executableSkills` filter drops `disableModelInvocation: true` docs from `activeSkillIds` whenever an invocable doc with the same name is in scope — `getSkillByName`'s `updatedAt desc` tiebreak can no longer pick the disabled doc and shadow the cataloged one. When ONLY a disabled doc exists for a name (no invocable counterpart), it stays in `activeSkillIds` so the explicit-rejection error path still fires for hallucinated invocations. Two new tests cover both cases.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/skills.ts [hunk 5]
```diff
@@ -213,15 +238,55 @@ export async function injectSkillCatalog(
...
+   * existing duplicate-name warn log below covers that ambiguity, and
+   * `getSkillByName` picks the newest (deterministic).
+   */
+  const invocableNames = new Set<string>();
+  for (const s of catalogVisibleSkills) {
+    invocableNames.add(s.name);
+  }
+  const executableSkills = activeSkills.filter(
+    (s) => s.disableModelInvocation !== true || !invocableNames.has(s.name),
+  );
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts:276]: Addressed in cc8d82045. `handleReadFileCall` now applies the same explicit-rejection gate as the skill tool — `disable-model-invocation: true` skills cannot have their SKILL.md body or bundled files read via `read_file` either. The shared `accessibleSkillIds` set still includes disabled skills (so the rejection path can fire instead of a misleading generic "not found"); each handler enforces its own contract. Two new tests cover the gate and regression-protect the happy path.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 6]
```diff
@@ -177,6 +184,24 @@ async function handleReadFileCall(
...
+  /**
+   * `disable-model-invocation: true` skills are off-limits for any model
+   * tool, not just the `skill` tool. Without this gate, a model that
+   * learned a disabled skill's name (stale catalog, hallucination) could
+   * still read its `SKILL.md` body or bundled files via `read_file`,
+   * defeating the contract. The skill stays in `accessibleSkillIds` only
+   * so handlers can fetch the doc and emit this explicit rejection
+   * instead of a misleading generic "not found".
+   */
+  if (skill.disableModelInvocation === true) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Addressed in 0b00cc78e. The `read_file` gate now bypasses for skills the user manually invoked this turn — `manualSkillNames` is plumbed from `manualSkillPrimes` through `agentToolContexts` → `enrichWithSkillConfigurable` → `mergedConfigurable`. Manually-primed disabled skills can now read their bundled `references/*` and `scripts/*` files. The autonomous-block contract is preserved (the bypass is name-scoped to the per-turn allowlist; a different disabled skill the user didn't pick still gets rejected). Two new tests cover both halves. Sticky-primed disabled skills from PRIOR turns are not yet in the allowlist — same-turn manual invocation was the load-bearing path; the prior-turn case is a known limitation in the commit message.

File: packages/api/src/agents/skills.ts [hunk 7]
```diff
@@ -414,6 +485,24 @@ export async function resolveManualSkills(
...
+         * "not found / inactive / empty body" pattern — a single misfiring
+         * skill must never block the user's actual message from going out.
+         *
+         * Checked before the body check because `userInvocable` is an
+         * authoritative author decision, while empty body could be a
+         * transient state — surfacing the more specific cause helps
+         * operators triage faster.
+         */
+        if (skill.userInvocable === false) {
+          logger.warn(`[resolveManualSkills] Skill "${name}" is not user-invocable — skipping`);
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts:542]: Addressed in 0b00cc78e. New `backfillDerivedFromFrontmatter` helper applied at read time in `getSkillByName` and `listSkillsByAccess` — when the structured column is unset but `frontmatter['user-invocable']` / `frontmatter['disable-model-invocation']` / `frontmatter['allowed-tools']` is present, the column gets populated on the returned object so all four runtime gates (catalog, popover, skill tool, read_file) fire correctly. `updateSkill` already re-derives, so legacy skills self-heal on next save. Column wins when both are set; backfill only fills undefined fields. Three new spec tests cover both methods and the column-wins precedence.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/handlers.ts [hunk 8]
```diff
@@ -177,6 +184,31 @@ async function handleReadFileCall(
...
+   * blocked. Bypass the gate for manually-primed skill names so manual `$`
+   * invocation of disabled skills stays usable end-to-end.
+   *
+   * Sticky-primed skills (manually or model-invoked in prior turns) are not
+   * yet in this exception list — that's a known limitation tracked for
+   * a follow-up. Same-turn manual invocation is the load-bearing path.
+   */
+  const manualSkillNames = (mergedConfigurable?.manualSkillNames as string[] | undefined) ?? [];
+  const isManuallyPrimedThisTurn = manualSkillNames.includes(skillName);
+  if (skill.disableModelInvocation === true && !isManuallyPrimedThisTurn) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Addressed in 530cf6f36. The OpenAI Responses (`responses.js`) and OpenAI completions (`openai.js`) controllers also call `enrichWithSkillConfigurable` and were ignoring the new `manualSkillNames` argument — both now pass `primaryConfig.manualSkillPrimes?.map(p => p.name)` so the manual-prime exception in the read_file gate works on every endpoint, not just the agents endpoint.

File: packages/api/src/agents/skills.ts [hunk 9]
```diff
@@ -213,15 +238,55 @@ export async function injectSkillCatalog(
...
+  }
+  const executableSkills = activeSkills.filter(
+    (s) => s.disableModelInvocation !== true || !invocableNames.has(s.name),
+  );
+
+  if (catalogVisibleSkills.length === 0) {
+    return {
+      toolDefinitions: inputDefs,
+      skillCount: 0,
+      activeSkillIds: executableSkills.map((s) => s._id),
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts]: Addressed in 530cf6f36. `injectSkillCatalog` no longer short-circuits all tool registration when `catalogVisibleSkills.length === 0`. Catalog text and the `skill` tool are still gated on having model-visible targets (no point burning tokens otherwise), but `read_file` and `bash_tool` now register whenever any active skill is in scope. Manually-primed disabled skills can still call read_file for their bundled `references/*` and `scripts/*` even in the all-disabled-only edge case. New test asserts the read_file-registered / skill-omitted shape; companion test verifies bash_tool joins when codeEnvAvailable.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/skills.ts [hunk 10]
```diff
@@ -414,6 +497,24 @@ export async function resolveManualSkills(
...
+         * skill must never block the user's actual message from going out.
+         *
+         * Checked before the body check because `userInvocable` is an
+         * authoritative author decision, while empty body could be a
+         * transient state — surfacing the more specific cause helps
+         * operators triage faster.
+         */
+        if (skill.userInvocable === false) {
+          logger.warn(`[resolveManualSkills] Skill "${name}" is not user-invocable — skipping`);
+          return null;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts:543]: Addressed in 09a20d146. `getSkillByName` gained an `options.preferInvocable` flag: when true, it prefers the newest doc that's both `userInvocable !== false` AND `disableModelInvocation !== true` (with frontmatter backfill), and falls back to the newest match otherwise. `resolveManualSkills` now passes `preferInvocable: true` so an older user-invocable doc is correctly resolved when a newer non-user-invocable duplicate would otherwise shadow it. The popover-visible skill matches what the resolver returns. New spec test covers both the collision case and the sole-disabled fallback.

File: packages/api/src/agents/skills.ts [hunk 11]
```diff
@@ -213,15 +238,47 @@ export async function injectSkillCatalog(
...
+   * existing duplicate-name warn log below covers that ambiguity, and
+   * `getSkillByName` picks the newest (deterministic).
+   */
+  const invocableNames = new Set<string>();
+  for (const s of catalogVisibleSkills) {
+    invocableNames.add(s.name);
+  }
+  const executableSkills = activeSkills.filter(
+    (s) => s.disableModelInvocation !== true || !invocableNames.has(s.name),
+  );
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts:276]: Addressed in 09a20d146. Same `preferInvocable` flag is also passed by `handleSkillToolCall` and `handleReadFileCall`, so the runtime ACL resolves to the catalog-visible doc rather than a newer disabled duplicate. With both manual prime and runtime handlers using `preferInvocable: true`, the same SKILL.md body the user / catalog targeted is the one whose files get read. The disable-model-invocation gate still fires correctly in the sole-disabled-name case (the fallback returns the disabled doc; the gate produces the explicit error).

**Comment by danny-avila** [comment]
@codex review

File: packages/data-schemas/src/methods/skill.ts [hunk 12]
```diff
@@ -617,13 +710,49 @@ export function createSkillMethods(mongoose: typeof import('mongoose'), deps: Sk
...
+      .lean()) as unknown as Array<ISkill & { _id: Types.ObjectId }>;
+    if (docs.length === 0) {
+      return null;
+    }
+    for (const doc of docs) {
+      backfillDerivedFromFrontmatter(doc);
+    }
+    const cleanInvocable = docs.find(
+      (d) => d.userInvocable !== false && d.disableModelInvocation !== true,
+    );
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/data-schemas/src/methods/skill.ts]: Addressed in 45d5c3817. Split `preferInvocable` into two independent axes: `preferUserInvocable` (filters on `userInvocable !== false`) and `preferModelInvocable` (filters on `disableModelInvocation !== true`). Manual paths now pass only `preferUserInvocable`; model paths (`handleSkillToolCall` / `handleReadFileCall`) pass only `preferModelInvocable`. Model-only skills (`userInvocable: false`) remain valid model-invocation targets — no longer shadowed by older user-invocable duplicates. Tests assert both that the right flag is passed and that the wrong flag is NOT passed (so manual-prime of disabled skills still works per iter 4).

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/initialize.ts [hunk 13]
```diff
@@ -360,6 +394,87 @@ export async function initializeAgent(
...
+      tools,
+      model: agent.model,
+      tool_options: agent.tool_options,
+      tool_resources,
+    });
+
+  let loadToolsResult;
+  try {
+    loadToolsResult = await callLoadTools(requestedToolNames);
+  } catch (err) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/initialize.ts:476]: Addressed in 3cb9eadf0. Added an explicit undefined-result branch alongside the throw branch — when `loadTools` returns `undefined` AND extras were requested, we retry with just `agent.tools`. Mirrors the swallowed-error production behavior of `createToolLoader` so the agent's baseline tools always load even when a skill-added tool silently fails. New test pins this behavior.

File: packages/api/src/agents/handlers.ts [hunk 14]
```diff
@@ -167,7 +183,15 @@ async function handleReadFileCall(
...
-  const skill = await getSkillByName(skillName, accessibleIds);
+  /* `preferModelInvocable` keeps name-collision resolution aligned with
+     the catalog: a newer `disable-model-invocation: true` duplicate
+     can't shadow the cataloged invocable doc the model actually
+     targeted. We don't filter by `userInvocable` here — model-only
+     skills (`userInvocable: false`) are valid model-invocation targets.
+     Falls back to the newest match so the disabled-only case still
+     resolves and the model-invocation gate below fires its explicit
+     error. */
+  const skill = await getSkillByName(skillName, accessibleIds, { preferModelInvocable: true });
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/handlers.ts]: Addressed in 3cb9eadf0. `handleReadFileCall` now picks its lookup option based on whether the skill is in this turn's `manualSkillNames`: `preferUserInvocable` for manually-primed skills (matches `resolveManualSkills`'s lookup so files come from the same doc whose body was primed), `preferModelInvocable` for autonomous probes (matches the catalog the model saw). Same-name collisions no longer cause body-vs-files mismatch on the manual path. Two new tests cover both branches.

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/agents/skills.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/skills.ts:276]: Addressed in aea9f11d1. `ResolvedManualSkill` now carries the primed doc's `_id`, and the legacy controllers build a `manualSkillPrimedIdsByName: Record<string, string>` map that flows through `enrichWithSkillConfigurable` to `mergedConfigurable`. `handleReadFileCall` now pins `accessibleIds` to `[primedId]` whenever the requested skill is in that map — the lookup returns the EXACT doc the resolver primed, regardless of how the dedup in `activeSkillIds` shaped the runtime ACL set. Body / files come from the same source even when same-name duplicates exist. Autonomous probes (skill not manually primed) keep the full ACL set + `preferModelInvocable`. New test covers the pinned-id lookup; existing tests updated for the new map shape.

**Comment by danny-avila** [comment]
@codex review
